task-maker-async/src/server.rs

```diff
@@ -23,3 +31,177 @@ pub trait Server {
+    options: ExecutionDAGOptions,
+    handle: FileSetWriteHandle,
+}
+
+impl Ord for WorkerTask {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Inverted order, so that tasks with lowest niceness are preferred, and on equality tasks
+        // that became ready first are preferred.
+        other.scheduling.cmp(&self.scheduling)
+    }
+}
+
+impl PartialOrd for WorkerTask {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+struct ServerImpl {
+    waiting_workers: Vec<Sender<()>>,
+    queue: BinaryHeap<WorkerTask>,
+    store: StoreService,
+}
+
+#[derive(Clone)]
+pub struct ServerService {
+    service: Arc<Mutex<ServerImpl>>,
+    // TODO(veluca): probably needs a CancellationToken too.
+}
+
+async fn wait_for_dependency(store: StoreService, hash: FileSetHash) -> Result<(), Error> {
+    store
+        .wait_for_fileset(
+            context::current(),
+            hash,
+            crate::store::WaitFor::Finalization,
+        )
+        .await
+}
+
+impl ServerService {
+    async fn evaluate_execution_group(
+        self,
+        execution: ExecutionGroup,
+        options: ExecutionDAGOptions,
+    ) -> Result<(), Error> {
+        let hash = FileSetHash {
+            data: execution.get_data_identification_hash(),
+            variant: execution.get_variant_identification_hash(),
+        };
+        let dependencies: Vec<_> = execution
+            .executions
+            .iter()
+            .flat_map(|ex| ex.files.iter())
+            .flat_map(|(_, f)| {
+                match f {
+                    ExecutionFileMode::Input(info) => Some(info.hash),
+                    _ => None,
+                }
+                .into_iter()
+            })
+            .collect();
+
+        // TODO(veluca): check for compatible variants.
+
+        let store = {
+            let service = self.service.lock().unwrap();
+            service.store.clone()
+        };
+
+        try_join_all(
+            dependencies
+                .iter()
+                .map(|hash| wait_for_dependency(store.clone(), *hash)),
+        )
+        .await?;
+
+        for _ in 0..MAX_RETRIES {
+            let store = store.clone();
+            let computation = store.create_computation(hash)?;
+            if computation.is_none() {
+                return Ok(());
+            }
+            let computation = computation.unwrap();
+
+            let task = WorkerTask {
+                scheduling: (options.niceness, execution.niceness, Instant::now()),
+                execution: execution.clone(),
+                options,
+                handle: computation,
+            };
+
+            {
+                let mut service = self.service.lock().unwrap();
+                service.queue.push(task);
+                // Wake up all the workers that are currently waiting for a new task.
+                service.waiting_workers.drain(..).for_each(|sender| {
```

Comment:
Why wake up everyone instead of just one?

Reply:
Because that worker might have disconnected in the meantime, which could cause a deadlock.

Of course, the best solution would be to just wake up the first one that is in fact still waiting, but since I think the worker count is going to be <= 100 in most (all?) situations, this ought to be OK as is...

Reply:
This mean that we will wake the workers (num workers * num executions) times. For now it may be fine, so I'd add a TODO for it.

To fix the deadlock one may simply track the execution assigned to a worker and reschedule it if it dies (the current version does this).

Reply:
Not necessarily (to both points):
- num_w * num_exec is just the worst case - if every (execution is ready) event is perfectly interleaved with all (the workers are woken up) events. In practice, I expect this worst case to not come close to being realized. I added a comment about it anyway.
- Note that waking up a worker doesn't necessarily mean that the task is assigned to that worker. Anyway there already is a retry logic.